pkg/storage/s3.go

```diff
@@ -1,56 +1,318 @@
+	flags.String(s3RegionOption, "", "Set the S3 region")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs")
+	flags.String(s3ProviderOption, "", "Set the S3 provider")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	options.SendCredential, err = flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
 	options.Endpoint, err = flags.GetString(s3EndpointOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	options.Region, err = flags.GetString(s3RegionOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
+	options.SSE, err = flags.GetString(s3SSEOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ACL, err = flags.GetString(s3ACLOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.StorageClass, err = flags.GetString(s3StorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ForcePathStyle = true
+	options.Provider, err = flags.GetString(s3ProviderOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	return options, err
+}
+
+// newS3Storage initialize a new s3 storage for metadata
+func newS3Storage(s3Back *backup.S3) (*S3Storage, error) {
+	qs := *s3Back
+	v := credentials.Value{
+		AccessKeyID:     qs.AccessKey,
+		SecretAccessKey: qs.SecretAccessKey,
+	}
+
+	// low timeout to ec2 metadata service
+	lowTimeoutClient := &http.Client{Timeout: 1 * time.Second}
+	def := defaults.Get()
+	def.Config.HTTPClient = lowTimeoutClient
+	ec2Session, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	// first provider to supply a credential set "wins"
+	providers := []credentials.Provider{
+		// use static credentials if they're present (checked by provider)
+		&credentials.StaticProvider{Value: v},
+
+		// * Access Key ID:     AWS_ACCESS_KEY_ID or AWS_ACCESS_KEY
+		// * Secret Access Key: AWS_SECRET_ACCESS_KEY or AWS_SECRET_KEY
+		&credentials.EnvProvider{},
+
+		// A SharedCredentialsProvider retrieves credentials
+		// from the current user's home directory.  It checks
+		// AWS_SHARED_CREDENTIALS_FILE and AWS_PROFILE too.
+		&credentials.SharedCredentialsProvider{},
+
+		// Pick up IAM role if we're in an ECS task
+		defaults.RemoteCredProvider(*def.Config, def.Handlers),
+
+		// Pick up IAM role in case we're on EC2
+		&ec2rolecreds.EC2RoleProvider{
+			Client: ec2metadata.New(ec2Session, &aws.Config{
+				HTTPClient: lowTimeoutClient,
+			}),
+			ExpiryWindow: 3 * time.Minute,
+		},
+	}
+	cred := credentials.NewChainCredentials(providers)
 
-	// TODO: Add more options here.
+	if qs.Region == "" && qs.Endpoint == "" {
+		qs.Endpoint = "https://s3.amazonaws.com/"
+	}
+	if qs.Region == "" {
+		qs.Region = "us-east-1"
+	}
```

Comment:
Let's move these into `apply()`

Reply:
done

Reply:
@DanielZhangQD isn't worse if Rust handle it differently? (e.g. BR uploaded stuff to `us-east-1` but Rust uploaded to some other random region)

Reply:
What I concern about is that `Region` is required by aws-sdk-go but may not in Rust lib.
If I send the updated region to TiKV, it may cause failure of Rust lib.
What about just move below part to `apply`:
```
if qs.Region == "" && qs.Endpoint == "" {
		qs.Endpoint = "https://s3.amazonaws.com/"
	}
```

Reply:
Will remove the whole block to `apply`.

Reply:
done